### ISSUE TITLE
Caching support

### ISSUE THREAD

**Issue by ogrodnek** [lambci/lambci#38]
It would be nice to have the option of caching one or more directories (to s3) in order to speed up build times.

I'd envision this working mostly like it does in Travis.  They even have a [great doc explaining how their caching works and the rules around it](https://docs.travis-ci.com/user/caching/#How-does-the-caching-work?).  They also just push/pull to s3 for their caching.

In the lamb config, you could specify which directories you want to cache, e.g.:

```
{
  cache: [
    "~/.iv2",
    "~/.local"
  ]
}
```

Following the same rules as outlined by travis, keep a cache per branch, use the master cache if branch cache doesn't exist.  Download the archives from s3.

After the build, make new archives, and figure out if you need to replace S3 artifacts (checksum/etags?)

Thinking can just use the same S3 bucket that's used for builds already...

I did a really really quick and dirty PoC as part of a build with some hardcoded paths, etc. for just ~/.ivy2 (sbt/ivy proj) and for even a really simple build (just 3 explicit dependencies), was able to drop the build time from 2 minutes to 1 minute.

[pull_cache](https://github.com/ogrodnek/args4j-helpers/blob/lamb/pull_cache) runs before build, installs awscli, pulls down cache archive

[cache_deps](https://github.com/ogrodnek/args4j-helpers/blob/lamb/cache_deps) runs after build, pushes back up.

[Build before cache](https://lambci-buildresults-cedw6hal5zcd.s3.amazonaws.com/gh/ogrodnek/args4j-helpers/builds/8/971557126d64839eb26226362fdbc070.html)

[Build after cache](https://lambci-buildresults-cedw6hal5zcd.s3.amazonaws.com/gh/ogrodnek/args4j-helpers/builds/24/a77ed5291d06e30bcbe186d679096402.html)

Also, like I said, this is just pretty quick and dirty... I would expect caching the local pip stuff (awscli), and also being smarter about when we need to push back to the cache would make this even faster... This is also a pretty small project... I'd expect more gains from more dependencies to a point...

would be interested to hear your thoughts around this.

thanks.

**Comment by mhart** [comment]
Yeah, I've been thinking about something similar – and actually about how to use it from a build that triggers another build (or builds) to address #2  – so you could have an install step, and then trigger multiple different test steps, and then finally have a deploy step (or something like that) – and you'd save the artifacts along the way.

My initial thought was just something where tasks in arrays happen in series and tasks in objects happen in parallel, like:
```
{
  cmd: [
    'npm install',
    {
      test: 'npm test',
      lint: 'npm run lint',
    },
    'npm run deploy',
  ]
}
```
So the first Lambda would run `npm install`, then when it was done, would spin up two new Lambdas in parallel to use the output of the first one (saved in S3), one to test and one to lint, and then when the last of those two were done, it would spin up a new Lambda to run the final deploy step.

Obviously with this you could specify a quite complex nested, parallel task structure – but the caching at each step is missing. Initially I had thought just to bundle up the entire project directory, and then untar it on the next execution – but perhaps it makes more sense to be a bit more explicit about it:

```
{
  cmd: [
    {
      cmd: 'npm install',
      output: [
        '~/.npm',
        '$LAMBCI_BUILD_DIR', // support env vars or something
      ],
    },
    {
      input: [ // could just be inferred from last step
        '~/.npm',
        '$LAMBCI_BUILD_DIR'
      ],
      cmd: {
        test: 'npm test',
        lint: 'npm run lint',
      },
    },
    'npm run deploy',
  ]
}
```
Not sure, will play around with it a bit. Anyway, slightly off topic.

Certainly in terms of explicit caching to speed up subsequent builds, I definitely want to implement this – and the way you've suggested looks good. Obviously having out-of-the-box support would be much cleaner, and wouldn't require the AWS CLI to be installed or anything.

I'll need to think about how to handle security – if private and public builds are being run together, you don't want a PR from a fork of the public build to get access to a private cache from another project – but that might already be covered by the current rules that prevent access to secret keys.

Thanks for opening! I'll let you know how I progress.

**Comment by ogrodnek** [comment]
The idea for pipelined builds with artifacts is pretty interesting!

Also makes me think that maybe build artifacts on their own are also interesting...

Glad you are planning on the explicit caching support!  Definitely out-of-the-box is much much cleaner.

Good point on thinking through the security...

From the [travis caching doc](https://docs.travis-ci.com/user/caching), it sounds like they keep a separate cache for the PR (but would be bootstrapped from the target branch or default cache)... (they also make a note of mentioning that [Cache content is accessible by PRs](https://docs.travis-ci.com/user/caching#Cache-content-can-be-accessed-by-pull-requests)

[Drone](https://github.com/drone/drone) takes a simpler approach and their [pull-requests only have read-only cache access](http://readme.drone.io/usage/caching/)

Definitely something to think through...

Was about to say if it's easier, maybe public PRs don't even have caching (although both the drone and travis approaches seem reasonable)... But then I was thinking, well, someone is running a build where you are footing the bill, so, might as well decrease it... which makes me think, kind of off-topic, but should there be public PR build throttling or something?
